File: gui/ve_approval_capability_check.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// VEApprovalCapabilityStatus represents the result of checking a VE's approval capability.
type VEApprovalCapabilityStatus struct {
	VEID          string `json:"ve_id"`
	HasCapability bool   `json:"has_capability"`
	Enabled       bool   `json:"enabled"`
	Error         string `json:"error,omitempty"`
}

// CheckVEApprovalCapability checks whether the specified VE has approval capability enabled.
// For the local VE (empty veID or matching the local machine ID), it reads from the local config.
// For remote VEs, it queries the hub for the VE's capability status.
//
// Returns (true, nil) if the VE has approval capability enabled.
// Returns (false, nil) if the VE exists but does not have approval capability enabled.
// Returns (false, error) if the VE cannot be found or an error occurs.
func (a *App) CheckVEApprovalCapability(veID string) (bool, error) {
	if veID == "" {
		return false, fmt.Errorf("ve_id is required")
	}

	// Check if this is the local VE by comparing with the configured machine ID.
	localMachineID := a.getLocalVEMachineID()
	if veID == localMachineID {
		return a.checkLocalVEApprovalCapability()
	}

	// For remote VEs, query the hub for capability status.
	return a.checkRemoteVEApprovalCapability(veID)
}

// CheckVEApprovalCapabilityStatus returns a detailed status object for the VE's approval capability.
// This is the Wails binding exposed to the frontend for use in the workflow designer's
// approval node configuration panel.
func (a *App) CheckVEApprovalCapabilityStatus(veID string) VEApprovalCapabilityStatus {
	if veID == "" {
		return VEApprovalCapabilityStatus{
			VEID:          veID,
			HasCapability: false,
			Enabled:       false,
			Error:         "ve_id is required",
		}
	}

	enabled, err := a.CheckVEApprovalCapability(veID)
	status := VEApprovalCapabilityStatus{
		VEID:          veID,
		HasCapability: enabled,
		Enabled:       enabled,
	}
	if err != nil {
		status.Error = err.Error()
	}
	return status
}

// ValidateVEApproverAssignment validates that a VE can be assigned as an approver
// in the workflow designer. It checks that the VE has approval capability enabled.
// Returns nil if the assignment is valid, or an error describing why it's invalid.
func (a *App) ValidateVEApproverAssignment(veID string) error {
	enabled, err := a.CheckVEApprovalCapability(veID)
	if err != nil {
		return fmt.Errorf("cannot verify VE approval capability: %w", err)
	}
	if !enabled {
		return fmt.Errorf("VE %q does not have approval capability enabled; please enable it in the VE's settings before assigning as approver", veID)
	}
	return nil
}

// checkLocalVEApprovalCapability checks the local VE's approval capability from config.
func (a *App) checkLocalVEApprovalCapability() (bool, error) {
	cfg, err := a.GetVEApprovalConfig()
	if err != nil {
		return false, fmt.Errorf("failed to load VE approval config: %w", err)
	}
	return cfg.Enabled, nil
}

// checkRemoteVEApprovalCapability queries the hub for a remote VE's approval capability.
// The hub maintains knowledge of connected VEs and their reported capabilities.
func (a *App) checkRemoteVEApprovalCapability(veID string) (bool, error) {
	cfg, err := a.LoadConfig()
	if err != nil {
		return false, fmt.Errorf("failed to load config: %w", err)
	}
	hubURL := strings.TrimSpace(cfg.RemoteHubURL)
	if hubURL == "" {
		return false, fmt.Errorf("hub URL not configured; cannot check remote VE capability")
	}

	endpoint := strings.TrimRight(hubURL, "/") + "/api/v1/ve/" + veID + "/approval_capability"

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return false, fmt.Errorf("failed to create request: %w", err)
	}

	// Add authentication token if available.
	if cfg.RemoteMachineToken != "" {
		req.Header.Set("Authorization", "Bearer "+cfg.RemoteMachineToken)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return false, fmt.Errorf("failed to query hub: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return false, fmt.Errorf("VE %q not found on hub", veID)
	}
	if resp.StatusCode != http.StatusOK {
		return false, fmt.Errorf("hub returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, fmt.Errorf("failed to read hub response: %w", err)
	}

	var result struct {
		Enabled bool   `json:"enabled"`
		Error   string `json:"error,omitempty"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return false, fmt.Errorf("failed to parse hub response: %w", err)
	}
	if result.Error != "" {
		return false, fmt.Errorf("hub error: %s", result.Error)
	}
	return result.Enabled, nil
}

// getLocalVEMachineID returns the local machine's ID from config.
func (a *App) getLocalVEMachineID() string {
	cfg, err := a.LoadConfig()
	if err != nil {
		return ""
	}
	return cfg.RemoteMachineID
}

File: guiapp/frontend/src/utils/veApprovalCapabilityCheck.ts
/**
 * VE Approval Capability Check Utility
 *
 * Provides validation for assigning VEs as approvers in the workflow designer.
 * When a user selects a VE as an approver in the approval node configuration panel,
 * this utility checks whether that VE has the approval capability enabled.
 *
 * Requirement 2.6: WHEN the user assigns a VE as an approver and that VE does not
 * have approval capability enabled, THE Workflow_Designer SHALL prevent the assignment
 * and display an error message indicating that the selected VE lacks approval capability.
 */

import { CheckVEApprovalCapabilityStatus, ValidateVEApproverAssignment } from '../../wailsjs/go/main/App';

/**
 * Result of a VE approval capability check.
 */
export interface VEApprovalCapabilityResult {
  veId: string;
  hasCapability: boolean;
  enabled: boolean;
  error?: string;
}

/**
 * Validation result for a VE approver assignment.
 */
export interface VEApproverValidationResult {
  valid: boolean;
  errorMessage?: string;
}

/**
 * Checks whether a VE has approval capability enabled.
 * Used by the approval node configuration panel to validate VE selection.
 *
 * @param veId - The VE machine ID to check
 * @returns Promise resolving to the capability status
 */
export async function checkVEApprovalCapability(veId: string): Promise<VEApprovalCapabilityResult> {
  if (!veId || veId.trim() === '') {
    return {
      veId: veId || '',
      hasCapability: false,
      enabled: false,
      error: 'VE ID is required',
    };
  }

  try {
    const status = await CheckVEApprovalCapabilityStatus(veId);
    return {
      veId: status.ve_id || veId,
      hasCapability: status.has_capability || false,
      enabled: status.enabled || false,
      error: status.error || undefined,
    };
  } catch (err) {
    return {
      veId,
      hasCapability: false,
      enabled: false,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

/**
 * Validates that a VE can be assigned as an approver in the workflow designer.
 * Returns a validation result indicating whether the assignment is valid.
 * If invalid, includes an error message suitable for display in the UI.
 *
 * @param veId - The VE machine ID to validate
 * @returns Promise resolving to the validation result
 */
export async function validateVEApproverAssignment(veId: string): Promise<VEApproverValidationResult> {
  if (!veId || veId.trim() === '') {
    return {
      valid: false,
      errorMessage: 'Please select a VE to assign as approver.',
    };
  }

  try {
    await ValidateVEApproverAssignment(veId);
    return { valid: true };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      valid: false,
      errorMessage: formatApprovalCapabilityError(message),
    };
  }
}

/**
 * Formats the error message from the backend into a user-friendly message
 * for display in the workflow designer's approval node configuration panel.
 */
function formatApprovalCapabilityError(backendError: string): string {
  if (backendError.includes('does not have approval capability enabled')) {
    return 'This VE does not have approval capability enabled. Please enable it in the VE\'s Approval Settings before assigning as approver.';
  }
  if (backendError.includes('not found')) {
    return 'The selected VE could not be found. It may be offline or not registered.';
  }
  if (backendError.includes('hub URL not configured')) {
    return 'Cannot verify VE capability: Hub connection is not configured.';
  }
  if (backendError.includes('hub client not available')) {
    return 'Cannot verify VE capability: Hub is not connected.';
  }
  // Fallback: return the backend error as-is.
  return `Cannot assign VE as approver: ${backendError}`;
}
